lib/api/apiUtils/object/abortMultipartUpload.js

```diff
@@ -93,29 +92,82 @@ function abortMultipartUpload(authInfo, bucketName, objectKey, uploadId, log,
                     skipDataDelete);
                 });
         },
-        function deleteObjectMetadata(mpuBucket, storedParts, destBucket, objectMD, skipDataDelete, next) {
-            if (!objectMD || metadataValMPUparams.uploadId !== objectMD.uploadId) {
+        // During Abort, we dynamically detect if the previous CompleteMPU call
+        // created potential object metadata wrongly, e.g. by creating
+        // an object version when some of the parts are missing.
+        // By passing a null objectMD, we tell the subsequent steps
+        // to skip the cleanup.
+        // Another approach is possible, but not supported by all backends:
+        // to honor the uploadId filter in standardMetadataValidateBucketAndObj
+        // ensuring the objMD returned has the right uploadId. But this is not
+        // supported by Metadata.
+        function ensureCleanupIsRequired(mpuBucket, storedParts, destBucket,
+            objectMD, skipDataDelete, next) {
+            if (!objectMD) {
+                return next(null, mpuBucket, storedParts, destBucket, null, skipDataDelete);
+            }
+
+            // If objectMD exists and has matching uploadId, use it directly
+            // This handles all non-versioned cases, and some versioned cases.
+            if (objectMD.uploadId === uploadId) {
                 return next(null, mpuBucket, storedParts, destBucket, objectMD, skipDataDelete);
             }
-            // In case there has been an error during cleanup after a complete MPU
-            // (e.g. failure to delete MPU MD in shadow bucket),
-            // we need to ensure that the MPU metadata is deleted.
-            log.debug('Object has existing metadata, deleting them', {
+
+            // If bucket is not versioned, no need to check versions:
+            // as the uploadId is not the same, we skip the cleanup.
+            if (!destBucket.isVersioningEnabled()) {
+                return next(null, mpuBucket, storedParts, destBucket, null, skipDataDelete);
+            }
+
+            // Otherwise, list all versions to find one with a matching uploadId.
+            return services.findObjectVersionByUploadId(bucketName, objectKey, uploadId, log, (err, foundVersion) => {
+                if (err) {
+                    log.warn('error finding object version by uploadId, proceeding without cleanup', {
+                        error: err,
+                        method: 'abortMultipartUpload.ensureCleanupIsRequired',
+                    });
+                    // On error, continue the abort without an objectMD to clean up.
+                    return next(null, mpuBucket, storedParts, destBucket, null, skipDataDelete);
+                }
+                return next(null, mpuBucket, storedParts, destBucket, foundVersion, skipDataDelete);
+            });
+        },
+        function deleteObjectMetadata(mpuBucket, storedParts, destBucket, objMDWithMatchingUploadID,
```

Comment:
nit: should we keep the old name of the parameter, to make the diff behave better (and see we are really just introducing an extra step):

```suggestion
        function deleteObjectMetadata(mpuBucket, storedParts, destBucket, objectMD,
```

(also the mention of `uploadID` later making things confusing, since we don't check it there : IMHO the step you are adding makes the code correct by "fixing" the `objectMD` where needed, so that throughout we can now trust the `objectMD`.... so the problem is abstracted a way from that point, and we don't need to make the code look more complicated)